### DISCUSSION TITLE
How to gather all validation_step_outputs at validation_epoch_end and run in rank_zero properly without deadlock?

### DISCUSSION THREAD

**Discussion by allanchan339** [Lightning-AI/pytorch-lightning#13041] [DDP / multi-GPU / multi-node] [answered]
Hi,

I am trying to gather all the output and label pairs in the validation epoch end to run a simple validation process.
First, all validation data will be separated into different devices (controlled by DDP).
The validation step is simple as follows:

```
def validation_step(self, batch, batch_idx):
        # use valid_metric
        feat, label = batch
        output = self.model(feat)
        del batch, feat
        torch.cuda.empty_cache()


        for i, metric in enumerate(self.metrics):
            metric.update(output[:,i], label[:,i])

        return {"label": label, "output": output}
```

To calculate an accurate metrics, I need to gather all outputs to a single device and log on rank 0 only as follows:

```
 def validation_epoch_end(self, validation_step_outputs):
        def _valid_epoch_end(validation_step_outputs):
            enable_Flag = False
            # AP refers to average_precision_score in torchmetrics
            all_labels = list(map(itemgetter('label'), validation_step_outputs))
            all_labels = torch.cat(all_labels).cpu().detach().numpy()
            all_outputs = list(map(itemgetter('output'), validation_step_outputs))
            all_outputs = torch.cat(all_outputs).cpu().detach().numpy()

            AP = []
            for i in range(1, 17+1):
                AP.append(np.nan_to_num(average_precision_score(all_labels
                                                                [:, i], all_outputs[:,  i])))
            mAP = np.mean(AP)
            tmp = EVENT_DICTIONARY_V2
            tmp = tmp.copy()
            for k, v in tmp.items():
                #dict["kick-off"] = AP[0]
                tmp[k] = AP[v]
            
            self.log('Valid/mAP', mAP, logger=True, prog_bar=True,
                     rank_zero_only=True if self.args.strategy != 'dp' and enable_Flag else False)

            label_cls = (list(EVENT_DICTIONARY_V2.keys()))
            zip_iterator = zip(label_cls, AP)
            AP_dictionary = dict(zip_iterator)
            self.log('Valid/AP', AP_dictionary, logger=True, prog_bar=False,
                     rank_zero_only=True if self.args.strategy != 'dp' and enable_Flag else False)

        _valid_epoch_end(validation_step_outputs)
```
The program works fine. However, the metric result is always different from DP. Therefore, I need to gather all data first.
However, the program doesn't even pass the validation step as the deadlock. Any good idea to modify this program so that I can gather

**Comment by rohitgr7** [comment]
with DP, you need to configure the metric computation differently
https://pytorch-lightning.readthedocs.io/en/stable/common/lightning_module.html#validating-with-dataparallel

with DDP to gather all outputs you can use `self.all_gather` under the condition of rank_zero

```
if self.trainer.is_global_zero:
    all_val_outs = self.all_gather(...)

**Reply by allanchan339** [reply]
I have tried to use `is_global_zero` flag, `dist_sync = False, rank_zero_only = True`, but deadlock appear. 

At last I keep the all_gather() in validation epoch end, while keep the sync on. This time the deadlock didn't come out. It is weird as two process calculate the metric twice. 

The new code as follow:

```
    def validation_epoch_end(self, validation_step_outputs):
        def _valid_epoch_end_ddp(validation_step_outputs):

            enable_Flag = True
            # AP refers to average_precision_score in torchmetrics
            all_labels = list(
                map(itemgetter('label'), validation_step_outputs))

            all_outputs = list(
                map(itemgetter('output'), validation_step_outputs))
            all_labels_tmp = list()
            all_outputs_tmp = list()
            for i in range(len(all_labels)):
                all_labels_tmp.append(
                    rearrange(all_labels[i], "g b c -> (g b) c"))
                all_outputs_tmp.append(
                    rearrange(all_outputs[i], "g b c -> (g b) c"))

            all_labels = torch.cat(all_labels_tmp).cpu().detach().numpy()
            all_outputs = torch.cat(
                all_outputs_tmp).cpu().detach().numpy()

            AP = []
            for i in range(1, 17+1):
                AP.append(np.nan_to_num(average_precision_score(all_labels
                                                                [:, i], all_outputs[:,  i])))
            mAP = np.mean(AP)
            tmp = EVENT_DICTIONARY_V2
            tmp = tmp.copy()
            for k, v in tmp.items():
                #dict["kick-off"] = AP[0]
                tmp[k] = AP[v]

            self.log('Valid/mAP', mAP, logger=True, prog_bar=True,
                     )

            label_cls = (list(EVENT_DICTIONARY_V2.keys()))
            zip_iterator = zip(label_cls, AP)
            AP_dictionary = dict(zip_iterator)
            self.log('Valid/AP', AP_dictionary, logger=True, prog_bar=False,
                     )

        if self.trainer.strategy.strategy_name == 'ddp':
            validation_step_outputs = self.all_gather(validation_step_outputs)

            _valid_epoch_end_ddp(validation_step_outputs)
```

I wonder if there are any chance to modify such that the validation metric calculation can be done in a single process.

**Reply by rohitgr7** [reply]
I don't see is_global_zero used here:

it should be something like:
```py
validation_step_outputs = ...

if self.trainer.is_global_zero:
     validation_step_outputs = self.all_gather(validation_step_outputs)
```
is it possible for you to share a similar example using [BoringModel](https://colab.research.google.com/github/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/bug_report/bug_report_model.ipynb). I can help you fix what you need in a better way maybe.

**Reply by allanchan339** [reply]
Ooops. I should reply to the conversation here. 

Sure. Please refer to the following BoringModel. Unfortunately, the training doesn't even pass the sanity check due to unknown reason deadlock.... 
To run the following prog, you may need to import some additional modules eg `operator, einops, sklearn and numpy`, because of consistency concerns with validation_step and validation_epoch_end above. 
```
import os

import torch
from torch.utils.data import DataLoader, Dataset
from operator import itemgetter
from einops import rearrange
import numpy as np

from pytorch_lightning import LightningModule, Trainer
from sklearn.metrics import average_precision_score

EVENT_DICTIONARY_V2 = {"A": 0, "B": 1, "C": 2, "D": 3, "E": 4, "F": 5,
                       "G": 6, "H": 7, "I": 8, "J": 9, "K": 10,
                       "M": 11, "N": 12, "O": 13, "R": 14, "P": 15, "G": 16}


class RandomDataset(Dataset):
    def __init__(self, size, num_samples):
        self.len = num_samples
        self.data = torch.randn(num_samples, size)
        self.label = torch.round(torch.rand(num_samples, 18))

    def __getitem__(self, index):
        return self.data[index], self.label[index]

    def __len__(self):
        return self.len


num_samples = 10000


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Sequential(
            torch.nn.Linear(32, 18),
            torch.nn.Sigmoid())
        self.criterion = torch.nn.BCELoss()

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        x, y = batch
        y_hat = self.layer(x)
        loss = self.criterion(y_hat, y)
        self.log("train_loss", loss)
        return loss

    def validation_epoch_end(self, validation_step_outputs):
        if self.trainer.is_global_zero:
            validation_step_outputs = self.all_gather(validation_step_outputs)

            # AP refers to average_precision_score in torchmetrics
            all_labels = list(
                map(itemgetter('label'), validation_step_outputs))

            all_outputs = list(
                map(itemgetter('output'), validation_step_outputs))
            all_labels_tmp = list()
            all_outputs_tmp = list()
            for i in range(len(all_labels)):
                all_labels_tmp.append(
                    rearrange(all_labels[i], "g b c -> (g b) c"))
                all_outputs_tmp.append(
                    rearrange(all_outputs[i], "g b c -> (g b) c"))

            all_labels = torch.cat(all_labels_tmp).cpu().detach().numpy()
            all_outputs = torch.cat(
                all_outputs_tmp).cpu().detach().numpy()

            AP = []
            for i in range(1, 17+1):
                AP.append(np.nan_to_num(average_precision_score(all_labels
                                                                [:, i], all_outputs[:,  i])))
            mAP = np.mean(AP)
            tmp = EVENT_DICTIONARY_V2
            tmp = tmp.copy()
            for k, v in tmp.items():
                tmp[k] = AP[v]

            self.log('Valid/mAP', mAP, logger=True,
                    prog_bar=True, rank_zero_only=True)

            label_cls = (list(EVENT_DICTIONARY_V2.keys()))
            zip_iterator = zip(label_cls, AP)
            AP_dictionary = dict(zip_iterator)
            self.log('Valid/AP', AP_dictionary, logger=True, prog_bar=False, rank_zero_only=True
                    )

    def validation_step(self, batch, batch_idx):

        x, y = batch
        y_hat = self.layer(x)
        loss = self.criterion(y_hat, y)
        return {"label": y, "output": y_hat}

    def test_step(self, batch, batch_idx):
        x, y = batch
        y_hat = self.layer(x)
        loss = self.criterion(y_hat, y)
        self.log("test_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run():
    train_data = DataLoader(RandomDataset(32, num_samples), batch_size=2048)
    val_data = DataLoader(RandomDataset(32, num_samples), batch_size=2048)
    test_data = DataLoader(RandomDataset(32, num_samples), batch_size=2048)

    model = BoringModel()
    trainer = Trainer(
        accelerator='cpu',
        default_root_dir=os.getcwd(),
        min_epochs=10000,
        max_epochs=20000,
        enable_model_summary=False,
        devices=2,
        strategy='ddp',
    )
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data)
    # trainer.test(model, dataloaders=test_data)


run()

```

**Reply by rohitgr7** [reply]
okay.. looks like `if self.trainer.is_global_zero:` is not required and all_gather needs to be called on all the devices.

**Reply by allanchan339** [reply]
It is still weird to me to compute the same metric multiple times on different processes... 

Actually, I have encountered another odd outcome. During testing, I need each process in test_step to write results on JSON file (json.dump). Rank 0 did it correctly. However, other devices/processes may be stopped/interrupted before the writing file process is finished. At test_epoch_end, the JSON decoder will report a decoding error due to missing `}` in some result files generated by non-zero-rank process.   Currently, I just force the test_step `return 0` to ensure the process handles the next batch after all writing JSON procedures end. Any idea that may cause the error like this?

**Reply by rohitgr7** [reply]
I think once you have gathered the outputs from all the ranks, rest of the computation can be kept inside `is_global_zero` block to disable the computation on all the devices.

**Reply by allanchan339** [reply]
I tried... it seems the non-rank process will finish his process earlier and caused the deadlock. Is it possible to insert a barrier // synchronization step so that the non-rank process will wait for the global rank zero process in pytorch lightning??

**Reply by rohitgr7** [reply]
you can use `trainer.strategy.barrier()`

**Reply by allanchan339** [reply]
This sentence caused a deadlock on the last step in the test epoch............. leading to a non-stopping program...

**Reply by lioutasb** [reply]
Hi @allanchan339, did you figure out how to avoid deadlocks while computing metrics only on one device? I have the same issue.

**Reply by cnut1648** [reply]
Hi @lioutasb and @allanchan339 , have you all figured it out? I have exact same issue...

**Reply by semaphore-egg** [reply]
> Hi @lioutasb and @allanchan339 , have you all figured it out? I have exact same issue...

Use torch Metrics。

**Reply by xieweiyi** [reply]
having exact the same problem when using 'nccl' backend and trying to gather all steps results before reducing them in validation_epoch_end.

**Reply by Owen718** [reply]
having exact the same problem when using 'nccl' backend and trying to gather all steps results in validation_epoch_end.


`    def validation_epoch_end(self,outputs):
        if self.trainer.is_global_zero:
            outputs = self.all_gather(outputs)
            self.preds = []
            self.gts = []
            for i,(results,targets) in enumerate(outputs):
                self.preds.append(results)
                self.gts.append(targets)
        return self.preds,self.gts
`

**Reply by bimtauer** [reply]
Same issue here, have a metric that needs to be computed across entire dataset. Branching with `if self.trainer.is_global_zero` leads to deadlock in the other ranks. In my case EarlyStopping on that process complains that it's tracking metric is not available. This also begs the questions why early stopping is triggered at all in each rank - shouldn't it be a global check?

**Reply by ecolss** [reply]
Same issue here, `is_global_zero` and `all_gather` make the code hang forever.

**Reply by furkanbiten** [reply]
> Same issue here, have a metric that needs to be computed across entire dataset. Branching with `if self.trainer.is_global_zero` leads to deadlock in the other ranks. In my case EarlyStopping on that process complains that it's tracking metric is not available. This also begs the questions why early stopping is triggered at all in each rank - shouldn't it be a global check?

Same happened for me in ModelCheckpoint.

**Reply by vgthengane** [reply]
anybody has solved this yet?

**Reply by Vishwas-Venkatachalapathy** [reply]
I am running into the same issue. Let me know if there is a fix for it

**Reply by akihironitta** [reply]
Just to clarify, the collective `all_gather` needs to be called in all processes but not only in the main process, otherwise it leads to hanging.

**Comment by allanchan339** [comment, marked as answer]
The solution is as follows:
Instead of using 
```
if self.trainer.is_global_zero:
    all_val_outs = self.all_gather(...)
```
The function above will hang as card 0 is trying to communicate other cards

You should write code in this way
```
all_val_out = self.all_gather(...)

if self.trainer.is_global_zero:
    # merge output and process

self.trainer.strategy.barrier() #to let other cards to wait

```

**Reply by kenanseyidov** [reply]
Can you please provide documentation which explains how data moves from card to card? Right now I struggle to understand what `all_gather`, `is_global_zero` and `barrier()` achieve.

**Reply by akihironitta** [reply]
For collective operations, such as all gather/reduce, these blog posts helped me understand how it works.
- https://towardsdatascience.com/visual-intuition-on-ring-allreduce-for-distributed-deep-learning-d1f34b4911da
- https://tech.preferred.jp/en/blog/technologies-behind-distributed-deep-learning-allreduce

`is_global_zero` is just a `bool` value that represents whether the process is running on the 0th device of all devices or not:
- https://lightning.ai/docs/pytorch/2.0.1/common/trainer.html#is-global-zero

`barrier()` makes the running process wait until other processes get to the same code line. Some references I find helpful are:
- https://pytorch.org/docs/2.0/distributed.html#torch.distributed.barrier
- https://discuss.pytorch.org/t/torch-distributed-barrier-used-in-multi-node-distributed-data-parallel-training/89711
- https://github.com/pytorch/pytorch/issues/54059
